pkg/check/node_perf.go

```diff
@@ -0,0 +1,193 @@
+	for _, perf := range perfList.Returnval {
+		counterIds = append(counterIds, perf.CounterId)
+		copyPerf := perf
+		counterIdPerfMetricIdMap[perf.CounterId] = &copyPerf
+	}
+
+	perfCounter := types.QueryPerfCounter{
+		This:      *ctx.VMClient.ServiceContent.PerfManager,
+		CounterId: counterIds,
+	}
+	perfCounters, err := methods.QueryPerfCounter(ctx.Context, ctx.VMClient.RoundTripper,&perfCounter)
+	if err != nil {
+		return err
+	}
+
+	for _, perfCounter := range perfCounters.Returnval {
+		if _, ok := counterIdPerfMetricIdMap[perfCounter.Key]; !ok {
+			continue
+		}
+		perfMetricIdMap[perfCounter.Name()] = counterIdPerfMetricIdMap[perfCounter.Key]
+	}
+	return nil
+}
+
+func (c *CheckNodeDiskPerf) GetPerfMetric(ctx *CheckContext, vm *mo.VirtualMachine, metricName string) (*types.PerfMetricId, error){
+	err := c.BuildCounterIdMap(ctx, vm)
+	if err != nil {
+		return nil, err
+	}
+	if _, ok := perfMetricIdMap[metricName]; !ok {
+		return nil, fmt.Errorf("unable to find metric with name[%s]", metricName)
+	}
+	return perfMetricIdMap[metricName], nil
+}
+
+func (c *CheckNodeDiskPerf) StartCheck() error {
+	return nil
+}
+
+func init() { }
+
+func (c *CheckNodeDiskPerf) PerformMetricCheck(ctx *CheckContext, node *v1.Node, vm *mo.VirtualMachine, mcd MetricCheckDef) error {
+
+	if _, ok := node.Labels["node-role.kubernetes.io/master"]; ok && mcd.checkMasters == false {
+		klog.V(2).Infof("skipping metric check of[%s] as it is not intended for master nodes", mcd.metricName)
+		return nil
+	} else if _, ok := node.Labels["node-role.kubernetes.io/worker"]; ok && mcd.checkWorkers == false {
+		klog.V(2).Infof("skipping metric check of[%s] as it is not intended for worker nodes", mcd.metricName)
+		return nil
+	}
+
+	var (
+		hasWarning bool = false
+		hasError bool = false
+	)
+
+	klog.V(2).Infof("performing metric check of: %s", mcd.metricName)
+
+	perfMetricId, err := c.GetPerfMetric(ctx, vm, mcd.metricName)
+	if err != nil {
+		klog.V(2).Infof("error getting perf metric id: %s", err.Error())
+		return nil
+	}
+
+	metricsSpec := []types.PerfQuerySpec{{
+		Entity: vm.ManagedEntity.Reference(),
+		MetricId: []types.PerfMetricId{*perfMetricId},
+		MaxSample: metricMaxSamples,
+		IntervalId: metricIntervalId,
+	}}
+
+	perfQueryReq := types.QueryPerf{
+		This: *ctx.VMClient.ServiceContent.PerfManager,
+		QuerySpec: metricsSpec,
+	}
+
+	perfQueryRes, err := methods.QueryPerf(ctx.Context, ctx.VMClient.RoundTripper, &perfQueryReq )
+	if err != nil {
+		klog.V(2).Infof("error querying perf metrics: %s", err.Error())
+		return nil
+	}
+
+	for _, queryResult := range perfQueryRes.Returnval {
+		basePerfMetricSeries := queryResult.(*types.PerfEntityMetric).Value
+		for _, perfMetricSeries := range basePerfMetricSeries {
+			metricSeries := perfMetricSeries.(*types.PerfMetricIntSeries)
+			klog.V(4).Infof("metric samples: %+v", metricSeries)
+			for _, value := range metricSeries.Value {
+				if value > mcd.errorLevel {
+					hasError = true
```

Comment:
It might be useful to log the metric value with `V(2)`. `V(4)` used above is not visible by default.